resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -110,117 +104,381 @@ private[spark] class Client(
+      .withPath("/v1/submissions/ping")
+      .withNewPort(SUBMISSION_SERVER_PORT_NAME)
+      .build()
+    kubernetesClient.pods().createNew()
+      .withNewMetadata()
+        .withName(kubernetesAppId)
+        .withLabels(driverKubernetesSelectors)
+        .endMetadata()
+      .withNewSpec()
+        .withRestartPolicy("OnFailure")
+        .addNewVolume()
+          .withName(SUBMISSION_APP_SECRET_VOLUME_NAME)
+          .withNewSecret()
+            .withSecretName(submitServerSecret.getMetadata.getName)
+            .endSecret()
+          .endVolume()
+        .addToVolumes(sslVolumes: _*)
+        .withServiceAccount(serviceAccount)
+        .addNewContainer()
+          .withName(DRIVER_CONTAINER_NAME)
+          .withImage(driverDockerImage)
+          .withImagePullPolicy("IfNotPresent")
+          .addNewVolumeMount()
+            .withName(SUBMISSION_APP_SECRET_VOLUME_NAME)
+            .withMountPath(secretDirectory)
+            .withReadOnly(true)
+            .endVolumeMount()
+          .addToVolumeMounts(sslVolumeMounts: _*)
+          .addNewEnv()
+            .withName(ENV_SUBMISSION_SECRET_LOCATION)
+            .withValue(s"$secretDirectory/$SUBMISSION_APP_SECRET_NAME")
+            .endEnv()
+          .addNewEnv()
+            .withName(ENV_SUBMISSION_SERVER_PORT)
+            .withValue(SUBMISSION_SERVER_PORT.toString)
+            .endEnv()
+          .addToEnv(sslEnvs: _*)
+          .withPorts(containerPorts.asJava)
+          .withNewReadinessProbe()
+            .withHttpGet(probePingHttpGet)
+            .endReadinessProbe()
+          .withNewLivenessProbe()
+            .withHttpGet(probePingHttpGet)
+            .withInitialDelaySeconds(DRIVER_INITIAL_LIVELINESS_CHECK_DELAY_SECONDS)
+            .endLivenessProbe()
+          .endContainer()
+        .endSpec()
+      .done()
+  }
+
+  private class DriverPodReadyWatcher(resolvedDriverPod: SettableFuture[Pod]) extends Watcher[Pod] {
+    override def eventReceived(action: Action, pod: Pod): Unit = {
+      if ((action == Action.ADDED || action == Action.MODIFIED)
+          && pod.getStatus.getPhase == "Running"
+          && !resolvedDriverPod.isDone) {
+        pod.getStatus
+          .getContainerStatuses
+          .asScala
+          .find(status =>
+            status.getName == DRIVER_CONTAINER_NAME && status.getReady)
+          .foreach { _ => resolvedDriverPod.set(pod) }
+      }
+    }
+
+    override def onClose(cause: KubernetesClientException): Unit = {
+      logDebug("Driver pod readiness watch closed.", cause)
+    }
+  }
+
+  private class DriverEndpointsReadyWatcher(resolvedDriverEndpoints: SettableFuture[Endpoints])
+      extends Watcher[Endpoints] {
+    override def eventReceived(action: Action, endpoints: Endpoints): Unit = {
+      if ((action == Action.ADDED) || (action == Action.MODIFIED)
+          && endpoints.getSubsets.asScala.nonEmpty
+          && endpoints.getSubsets.asScala.exists(_.getAddresses.asScala.nonEmpty)
```

Comment:
On a second thought, we could still keep the pod watcher, such that we can tell the user the exact stage at which the submission fails (failed when creating the driver pod v.s. failed after the driver pod has been created)